### PR TITLE
ARROW-520: [C++] STL-compliant allocator

### PR DESCRIPTION
**Pull request by JohanMabille**
Ready for review

### CODE REVIEW

**Comment by wesm** [comment]
Thanks @JohanMabille! Would you mind running clang-format (`make format` from CMake) on the patch? I'll review in more detail tomorrow

**Review by xhochy — APPROVED** [review]
+1, this is really nice!

File: cpp/src/arrow/allocator-test.cc [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  stl_allocator<uint64_t> alloc;
+  uint64_t* data = alloc.allocate(100);
+
+  ASSERT_EQ(100 * sizeof(uint64_t), pool->bytes_allocated());
+
+  alloc.deallocate(data, 100);
+  ASSERT_EQ(0, pool->bytes_allocated());
+}
+
+TEST(stl_allocator, TestOOM) {
```
Comment (wesm) [inline cpp/src/arrow/allocator-test.cc]: Looks like this is failing in Valgrind -- move this test case into the if block below?

1: [ RUN      ] stl_allocator.TestOOM
1: ==10654== Argument 'size' of function memalign has a fishy (possibly negative) value: -8
1: ==10654==    at 0x4C2D110: memalign (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
1: ==10654==    by 0x4C2D227: posix_memalign (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
1: ==10654==    by 0x446C95: arrow::(anonymous namespace)::AllocateAligned(long, unsigned char**) (memory_pool.cc:48)
1: ==10654==    by 0x446FBF: arrow::DefaultMemoryPool::Allocate(long, unsigned char**) (memory_pool.cc:78)
1: ==10654==    by 0x442245: arrow::stl_allocator<unsigned long>::allocate(unsigned long, void const*) (allocator.h:59)
1: ==10654==    by 0x441533: arrow::stl_allocator_TestOOM_Test::TestBody() (allocator-test.cc:39)
1: ==10654==    by 0x46AFF7: void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (gtest.cc:2078)
1: ==10654==    by 0x466791: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (gtest.cc:2114)
1: ==10654==    by 0x44DD7A: testing::Test::Run() (gtest.cc:2151)
1: ==10654==    by 0x44E508: testing::TestInfo::Run() (gtest.cc:2326)
1: ==10654==    by 0x44EB43: testing::TestCase::Run() (gtest.cc:2444)
1: ==10654==    by 0x4552FE: testing::internal::UnitTestImpl::RunAllTests() (gtest.cc:4315)
1: ==10654== 
```

File: cpp/src/arrow/allocator.h [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef ARROW_UTIL_ALLOCATOR_H
+#define ARROW_UTIL_ALLOCATOR_H
```
Comment (wesm) [inline cpp/src/arrow/allocator.h]: The include guards don't match the location of the header

**Review by wesm — COMMENTED** [review]
There are some cpplint errors:

```
Done processing /home/travis/build/apache/arrow/cpp/src/arrow/memory_pool.h
/home/travis/build/apache/arrow/cpp/src/arrow/allocator.h:24:  Include the directory when naming .h files  [build/include] [4]
/home/travis/build/apache/arrow/cpp/src/arrow/allocator.h:25:  Include the directory when naming .h files  [build/include] [4]
/home/travis/build/apache/arrow/cpp/src/arrow/allocator.h:46:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]
/home/travis/build/apache/arrow/cpp/src/arrow/allocator.h:72:  Using C-style cast.  Use reinterpret_cast<void*>(...) instead  [readability/casting] [4]
```

You can run cpplint on the codebase with `make lint`

**Comment by JohanMabille** [comment]
I think I did something wrong with git, I had fixed them yesterday. I'll fix them again.
Anyway, is there any additional dependency required by lint ? running ``make lint`` complains with no info:
```bash
make[3]: *** [CMakeFiles/lint.dir/build.make:57: CMakeFiles/lint] Error 1
make[2]: *** [CMakeFiles/Makefile2:344: CMakeFiles/lint.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:351: CMakeFiles/lint.dir/rule] Error 2
make: *** [Makefile:277: lint] Error 2
```

**Comment by wesm** [comment]
cpplint is Python 2.7 only, you probably have 3.x in your path

**Comment by wesm** [comment]
+1, I'm merging now. thanks for your contribution! in the future, don't feel obligated to rebase as our PR merge tool will automatically squash all commits in the PR

**Pull request closed without merging**